oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/CachedGroupMembershipReader.java

```diff
@@ -0,0 +1,209 @@
+
+/**
+ * <code>CachedGroupMembershipReader</code>...
+ */
+class CachedGroupMembershipReader extends GroupMembershipReader {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CachedGroupMembershipReader.class);
+
+    private static final long MEMBERSHIP_THRESHOLD = 0;
+
+    /**
+     * Keep track of cache updates for 100 most recent authorizables.
+     */
+    private static final int MAX_CACHE_TRACKING_ENTRIES = 100;
+
+    private static final Map<UserConfiguration, Map<String, Long>> CACHE_UPDATES = new ConcurrentHashMap<>();
+
+    private final UserConfiguration config;
+
+    private final Root root;
+
+    private final long expiration;
+    private final long maxStale;
+
+    CachedGroupMembershipReader(@NotNull MembershipProvider membershipProvider,
+                                @NotNull GroupPrincipalFactory groupPrincipalFactory,
+                                @NotNull UserConfiguration config,
+                                @NotNull Root root) {
+        super(membershipProvider, groupPrincipalFactory);
+        this.config = checkNotNull(config);
+        this.root = root;
+        this.expiration = config.getParameters().getConfigValue(PARAM_CACHE_EXPIRATION, EXPIRATION_NO_CACHE);
+        this.maxStale = config.getParameters().getConfigValue(PARAM_CACHE_MAX_STALE, NO_STALE_CACHE);
+    }
+
+    @Override
+    void getMembership(@NotNull Tree authorizable,
+                       @NotNull Set<Principal> groupPrincipals) {
+        // membership cache only implemented on user
+        if (UserUtil.isType(authorizable, AuthorizableType.USER)) {
+            readGroupsFromCache(authorizable, groupPrincipals, this::loadGroupPrincipals);
+        } else {
+            loadGroupPrincipals(authorizable, groupPrincipals);
+        }
+    }
+
+    private void readGroupsFromCache(@NotNull Tree authorizableNode,
+                                     @NotNull Set<Principal> groups,
+                                     @NotNull BiConsumer<Tree, Set<Principal>> loader) {
+        long expirationTime = 0;
+        String authorizablePath = authorizableNode.getPath();
+        Tree principalCache = authorizableNode.getChild(CacheConstants.REP_CACHE);
+        if (principalCache.exists()) {
+            expirationTime = TreeUtil.getLong(principalCache, CacheConstants.REP_EXPIRATION, EXPIRATION_NO_CACHE);
+        }
+        long now = System.currentTimeMillis();
+        if (expirationTime > EXPIRATION_NO_CACHE && now < expirationTime) {
+            serveGroupsFromCache(authorizablePath, principalCache, groups);
+            return;
+        }
+
+        // need to load or serve stale if allowed
+        boolean mayServeStale = now - expirationTime < maxStale;
+        boolean updateCache;
+        Map<String, Long> updates = getCacheUpdateMap();
+        synchronized (updates) {
+            // anyone else updating that entry already?
```

Comment:
instead of asking a question, i would find the comment a lot more helpful if it would explain. "test if the cache for the given user is already being updated. this is achieved by ....." (please fill in .... to make it easily understandable for anyone looking at this code. at a first glance that was immediately obvious to me and how the exp time play a role here)